### PR TITLE
[CI] Binaries are available for macOS aarch64 and x86_64

### PR DESCRIPTION
**Pull request by maliasadi**
- [X] Add wheels recipes for macOS `aarch64` and `x86_64`
- [X] Update Make commands to support building from src on both Linux and macOS
- [X] Fix a few minor issues in the build system and typos

[sc-46388]

### CODE REVIEW

File: runtime/Makefile [hunk 1]
```diff
@@ -1,25 +1,33 @@
...
-ENABLE_ASAN?=OFF
+COMPILER_LAUNCHER?=$(shell which ccache)
+
 MK_ABSPATH := $(abspath $(lastword $(MAKEFILE_LIST)))
 MK_DIR := $(dir $(MK_ABSPATH))
 RT_BUILD_DIR?=$(MK_DIR)/build
 CMAKE_ARGS?=""
 CODE_COVERAGE?=OFF
-BUILD_TYPE?=Release
+BUILD_TYPE?=RelWithDebInfo
```
Comment (dime10) [inline runtime/Makefile:11]: Why change the default build type?

File: runtime/tests/CMakeLists.txt [hunk 2]
```diff
@@ -26,7 +26,7 @@ add_executable(runner_tests runner_main.cpp)
 
 add_subdirectory(third_party)
 
-if (ENABLE_ADDRESS_SANITIZER)
+if (ENABLE_ADDRESS_SANITIZER OR APPLE)
```
Comment (dime10) [inline runtime/tests/CMakeLists.txt:29]: Is this correct?

```suggestion
# Skip dynamic device loading test when sanitizers are enabled or when building on mac, due to
# RTLD_DEEPBIND incompatibility.
if (ENABLE_ADDRESS_SANITIZER OR APPLE)
```

File: runtime/tests/Test_LightningCoreQIS.cpp [hunk 3]
```diff
@@ -334,7 +334,9 @@ TEST_CASE("Test memref alloc", "[CoreQIS]")
 TEST_CASE("Test memref aligned alloc", "[CoreQIS]")
 {
     __quantum__rt__initialize();
-    int *a = (int *)_mlir_memref_to_llvm_aligned_alloc(sizeof(int), sizeof(int));
+
+    // macOS requires the alignment of 'aligned_alloc' is at least 8-byte
+    double *a = (double *)_mlir_memref_to_llvm_aligned_alloc(sizeof(double), sizeof(double));
```
Comment (dime10) [inline runtime/tests/Test_LightningCoreQIS.cpp:339]: Why were the types changed, shouldn't it work with integers still, just the alignment needs to be changed? Or is there something else going on?
```suggestion
    int *a = (int *)_mlir_memref_to_llvm_aligned_alloc(8, sizeof(int));
```

File: .github/workflows/build-wheel-macos-x86_64.yaml [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+    - name: Install PennyLane Plugins
+      run: |
+        python${{ matrix.python_version }} -m pip install PennyLane-Lightning-Kokkos
+        python${{ matrix.python_version }} -m pip install amazon-braket-pennylane-plugin
+
+    - name: Run Python Pytest Tests
+      run: |
+        python${{ matrix.python_version }} -m pytest $GITHUB_WORKSPACE/frontend/test/pytest -n auto
+        python${{ matrix.python_version }} -m pytest $GITHUB_WORKSPACE/frontend/test/pytest --runbraket=LOCAL -n auto
```
Comment (dime10) [inline .github/workflows/build-wheel-macos-x86_64.yaml:345]: Do we not support Kokkos here?

File: .github/workflows/build-wheel-macos-x86_64.yaml [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+# TODO: remove on pull_request before merging
+on:
+  pull_request:
+  push:
+    branches: [ main ]
+  workflow_dispatch:
+
+env:
+  MACOSX_DEPLOYMENT_TARGET: 11.7
```
Comment (dime10) [inline .github/workflows/build-wheel-macos-x86_64.yaml:11]: I guess apple stopped developing macos on X86? 😅

**Review by dime10 — COMMENTED** [review]
Great work @maliasadi 🥳 

I see the two actions are practically identical, should we not make use of the matrix feature?

File: .github/workflows/build-wheel-macos-x86_64.yaml (hunk 5 above)
Comment (maliasadi) [reply .github/workflows/build-wheel-macos-x86_64.yaml:11]: Hehe.. yeah these are minimum supported versions on these runners with backward compatibility support.

File: .github/workflows/build-wheel-macos-x86_64.yaml (hunk 4 above)
Comment (maliasadi) [reply .github/workflows/build-wheel-macos-x86_64.yaml:345]: I shall add this after fixing https://github.com/PennyLaneAI/pennylane-lightning/issues/552
```suggestion
        python${{ matrix.python_version }} -m pytest $GITHUB_WORKSPACE/frontend/test/pytest -n auto
        python${{ matrix.python_version }} -m pytest $GITHUB_WORKSPACE/frontend/test/pytest --backend="lightning.kokkos" -n auto
        python${{ matrix.python_version }} -m pytest $GITHUB_WORKSPACE/frontend/test/pytest --runbraket=LOCAL -n auto
```

File: mlir/Makefile [hunk 6]
```diff
@@ -1,19 +1,30 @@
-
 PYTHON?=python3
 C_COMPILER?=clang
 CXX_COMPILER?=clang++
-COMPILER_LAUNCHER?=ccache
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply mlir/Makefile:5]: Yep, on systems with no installation of `ccache`, we had to use `COMPILER_LAUNCHER="" make all` :( 

This simplifies the make command calls on different machines.

File: runtime/Makefile (hunk 1 above)
Comment (maliasadi) [reply runtime/Makefile:11]: I realized building with debug info is really a great help when it comes to debugging and fixing runtime-related bugs.

File: runtime/tests/CMakeLists.txt (hunk 2 above)
Comment (maliasadi) [reply runtime/tests/CMakeLists.txt:29]: I've updated your suggestion! I'll revisit the dynamic device loading work on macOS as the current support is not fully compatible with the macOS dylib API.

File: .github/workflows/build-wheel-macos-arm64.yaml [hunk 7]
```diff
@@ -0,0 +1,343 @@
+name: Build Catalyst Wheel on macOS (arm64)
+
+# TODO: remove on pull_request before merging
+on:
+  pull_request:
```
[the remark this answers is not in the document]
Comment (maliasadi) [reply .github/workflows/build-wheel-macos-arm64.yaml:5]: Let's remove these TODOs before merging the PR.

File: mlir/Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (dime10) [reply mlir/Makefile:5]: Oh nice, so `$(which ccache)` will produce an empty string when it doesn't exist? :)

File: .github/workflows/build-wheel-macos-arm64.yaml [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+    - name: Checkout Catalyst repo
+      uses: actions/checkout@v3
+
+    # Cache external project sources
+    - name: Cache LLVM Source
+      id: cache-llvm-source
+      uses: actions/cache@v3
+      with:
+        path:  mlir/llvm-project
+        key: ${{ runner.os }}-${{ runner.arch }}-llvm-${{ needs.constants.outputs.llvm_version }}-generic-source
```
[the remark this answers is not in the document]
Comment (dime10) [reply .github/workflows/build-wheel-macos-arm64.yaml:45]: This is I think:
```suggestion
        key: Linux-llvm-${{ needs.constants.outputs.llvm_version }}-generic-source
```

File: .github/workflows/build-wheel-macos-x86_64.yaml (hunk 4 above)
Comment (dime10) [reply .github/workflows/build-wheel-macos-x86_64.yaml]: Ah I see, could we add a comment / todo here? :)

File: Makefile [hunk 9]
```diff
@@ -1,4 +1,4 @@
-PYTHON ?= python3
+PYTHON ?= $(shell which python3)
```
Comment (dime10) [inline Makefile:1]: Good idea 👍

File: mlir/Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply mlir/Makefile:5]: yes, it does :)

File: runtime/Makefile (hunk 1 above)
Comment (maliasadi) [reply runtime/Makefile:11]: I remove it from the PR. Let's revisit this in a separate PR to check its overall impact on the wheels.

File: .github/workflows/build-wheel-macos-arm64.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (maliasadi) [reply .github/workflows/build-wheel-macos-arm64.yaml]: Great catch! :+1:

File: .github/workflows/build-wheel-macos-x86_64.yaml (hunk 4 above)
Comment (maliasadi) [reply .github/workflows/build-wheel-macos-x86_64.yaml]: here [2460cf1](https://github.com/PennyLaneAI/catalyst/pull/347/commits/2460cf1bbadd8e89ab016485d76f96471c13a366) :slightly_smiling_face:

File: .github/workflows/build-wheel-macos-x86_64.yaml [hunk 10]
```diff
@@ -0,0 +1,347 @@
+name: Build Catalyst Wheel on macOS (x86_64)
+
+# TODO: remove on pull_request before merging
+on:
+  pull_request:
```
Comment (maliasadi) [inline .github/workflows/build-wheel-macos-x86_64.yaml:5]:
```suggestion
on:
```

File: runtime/tests/CMakeLists.txt (hunk 2 above)
Comment (dime10) [reply runtime/tests/CMakeLists.txt:29]: Thanks!

**Review by dime10 — APPROVED** [review]
Great work 🦝

**Pull request merged**